traitlets/utils/decorators.py

```diff
@@ -0,0 +1,40 @@
+from .. import Undefined
+
+
+def enable_auto_completion(cls):
+    """Returns a decorator enabling auto-completion for the class traits"""
+
+    def format_value(value):
+        if value is Undefined:
+            return 'None'
+        if isinstance(value, str):
+            return '\"{}\"'.format(value)
+        return value
+
+    traits = [
+        (name, format_value(value.default_value))
+        for name, value in cls.class_traits().items()
+        if not name.startswith('_')
+    ]
+
+    formatting_vars = {
+        'cls_name': cls.__name__,
+        'args_def': ', '.join([
+            '{}={}'.format(name, value)
+            for name, value in traits
+        ])
+    }
+
+    code = """
+from inspect import signature
+
+def init_sig(self, *args, {args_def}, **kwargs):
+    pass
+
+{cls_name}.__init__.__signature__ = signature(init_sig)
```

Comment:
I think you can directly create signatures using `inspect.Signature` instead of having to dynamically defining a function in a string like this:

```python
from inspect import Signature, Parameter

sig = Signature([
    Parameter("x", kind=Parameter.KEYWORD_ONLY),
    Parameter("y", kind=Parameter.KEYWORD_ONLY),
])

print(sig)
```
```
(*, x, y)
```

This will also solve the problem of default values from your `format_value` function.

Reply:
Definitely :) And the exec can be removed too, I don't recall how/why I ended up writing an `exec` call with formatting. I'll clean that all. Thanks :)